pkg/trace/obfuscate/sql.go

```diff
@@ -142,36 +156,107 @@ func (f *groupingFilter) Filter(token, lastToken TokenKind, buffer []byte) (toke
 	f.groupFilter = 0
 	f.groupMulti = 0
 }
 
-// Process the given SQL or No-SQL string so that the resulting one is properly altered. This
-// function is generic and the behavior changes according to chosen tokenFilter implementations.
-// The process calls all filters inside the []tokenFilter.
-func (o *Obfuscator) obfuscateSQLString(in string) (string, error) {
-	literalEscapes := o.SQLLiteralEscapes()
-	tokenizer := NewSQLTokenizer(in, literalEscapes)
-	filters := []tokenFilter{&discardFilter{}, &replaceFilter{}, &groupingFilter{}}
-
-	out, err := attemptObfuscation(tokenizer, filters)
-	if err != nil && tokenizer.SeenEscape() {
+// obfuscateSQLString quantizes and obfuscates the given input SQL query string. Quantization removes
+// some elements such as comments and aliases and obfuscation attempts to hide sensitive information
+// in strings and numbers by redacting them.
+func (o *Obfuscator) obfuscateSQLString(in string) (*obfuscatedQuery, error) {
+	lesc := o.SQLLiteralEscapes()
+	tok := NewSQLTokenizer(in, lesc)
+	out, err := attemptObfuscation(tok)
+	if err != nil && tok.SeenEscape() {
 		// If the tokenizer failed, but saw an escape character in the process,
 		// try again treating escapes differently
-		tokenizer = NewSQLTokenizer(in, !literalEscapes)
-		if out, err2 := attemptObfuscation(tokenizer, filters); err2 == nil {
-			// If the second attempt succeeded, change the default behavior
-			o.SetSQLLiteralEscapes(!literalEscapes)
+		tok = NewSQLTokenizer(in, !lesc)
+		if out, err2 := attemptObfuscation(tok); err2 == nil {
+			// If the second attempt succeeded, change the default behavior so that
+			// on the next run we get it right in the first run.
+			o.SetSQLLiteralEscapes(!lesc)
 			return out, nil
 		}
 	}
 	return out, err
 }
 
-func attemptObfuscation(tokenizer *SQLTokenizer, filters []tokenFilter) (string, error) {
+// tableFinderFilter is a filter which attempts to identify the table name as it goes through each
+// token in a query.
+type tableFinderFilter struct {
+	// seen keeps track of unique table names encountered by the filter.
+	seen map[string]struct{}
+	// csv specifies a comma-separated list of tables
+	csv strings.Builder
+}
+
+// Filter implements tokenFilter.
+func (f *tableFinderFilter) Filter(token, lastToken TokenKind, buffer []byte) (TokenKind, []byte, error) {
+	switch lastToken {
+	case From:
+		// SELECT ... FROM [tableName]
+		// DELETE FROM [tableName]
+		if r, _ := utf8.DecodeRune(buffer); !unicode.IsLetter(r) {
+			// first character in buffer is not a letter; we might have a nested
+			// query like SELECT * FROM (SELECT ...)
+			break
+		}
+		fallthrough
+	case Update, Into, Join:
+		// UPDATE [tableName]
+		// INSERT INTO [tableName]
+		// ... JOIN [tableName]
+		f.storeName(string(buffer))
+	}
+	return token, buffer, nil
+}
+
+// storeName marks the given table name as seen in the internal storage.
+func (f *tableFinderFilter) storeName(name string) {
+	if _, ok := f.seen[name]; ok {
```

Comment:
Is if valid for a query to repeat the same tablename with different casing? We may not want to duplicate if that's the case. e.g. "SELECT TABLE.field FROM table" may produce "TABLE,table".

If we want to handle this case, I would recommend changing `seen` to a `map[string]string` and using a canonical key (lowercase?) in your `seen` map, and storing the first instance as the value so that we can maintain the original casing.

Reply:
It's a nit but you're totally right.